sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala

```diff
@@ -538,8 +538,12 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
       val cacheName = cache.get.cachedRepresentation.cacheBuilder.tableName
       val cacheLevel = cache.get.cachedRepresentation.cacheBuilder.storageLevel
 
+      // creates a new logical plan since the old table refers to old relation which
+      // should be refreshed
+      val newTable = sparkSession.table(tableIdent)
+
       // recache with the same name and cache level.
-      sparkSession.sharedState.cacheManager.cacheQuery(table, cacheName, cacheLevel)
+      sparkSession.sharedState.cacheManager.cacheQuery(newTable, cacheName, cacheLevel)
```

Comment:
BTW, I think we can just call `cacheTable(tableName, storageLevel)` API. It does the same thing.

Reply:
I think `cacheTable` takes a string as table name? we'll need to modify or have a new overloaded version that takes `TableIdentifier`.

Reply:
Passing the `tableName`?

Reply:
oops you're right, let me change that.

Reply:
@viirya even though `cacheTable` looks nicer, we may need to follow the old logic here because of table name. Someone can first cache a table using `db.tbl` as table name and then call `refreshTable` on `tbl` within the database `db`. In this case we'll need to pick up the `db` part from `cache.get.cachedRepresentation.cacheBuilder.tableName`.

Reply:
Hmm, I see. `cacheTable` doesn't take a custom cache table name as parameter so far.

Reply:
Yea, it also doesn't take `None` as table name but ppl could pass `None` when caching, and perhaps we should maintain that when recaching the table.

Reply:
Okay good. Let's go back to old logic. Thanks for the try.